pkg/aws/throttle/throttler_test.go

```diff
@@ -0,0 +1,199 @@
+		args            args
+		testDuration    time.Duration
+		testQPS         int64
+		validCallsCount func(elapsedDuration time.Duration, observedCallsCount int64)
+	}{
+		{
+			name: "single matching limiter",
+			fields: fields{
+				predictLimiters: []predictLimiter{
+					{
+						predict: func(r *request.Request) bool {
+							return true
+						},
+						limiter: rate.NewLimiter(10, 5),
+					},
+				},
+			},
+			args: args{
+				r: &request.Request{
+					HTTPRequest: &http.Request{},
+				},
+			},
+			testQPS: 100,
+			validCallsCount: func(elapsedDuration time.Duration, count int64) {
+				ideal := 5 + 10*elapsedDuration.Seconds()
+				// We should never get more requests than allowed.
+				if want := int64(ideal + 1); count > want {
+					t.Errorf("count = %d, want %d (ideal %f", count, want, ideal)
+				}
+				// We should get very close to the number of requests allowed.
+				if want := int64(ideal - 1); count < want {
+					t.Errorf("count = %d, want %d (ideal %f", count, want, ideal)
+				}
+			},
+		},
+		{
+			name: "single non-matching limiter",
+			fields: fields{
+				predictLimiters: []predictLimiter{
+					{
+						predict: func(r *request.Request) bool {
+							return false
+						},
+						limiter: rate.NewLimiter(10, 5),
+					},
+				},
+			},
+			args: args{
+				r: &request.Request{
+					HTTPRequest: &http.Request{},
+				},
+			},
+			testQPS: 100,
+			validCallsCount: func(elapsedDuration time.Duration, count int64) {
+				ideal := 100 * elapsedDuration.Seconds()
+				// We should never get more requests than allowed.
+				if want := int64(ideal + 1); count > want {
+					t.Errorf("count = %d, want %d (ideal %f", count, want, ideal)
+				}
+				// We should get very close to the number of requests allowed.
+				if want := int64(ideal - 1); count < want {
+					t.Errorf("count = %d, want %d (ideal %f", count, want, ideal)
+				}
+			},
+		},
+		{
+			name: "two limiter, one matching and another non-matching",
+			fields: fields{
+				predictLimiters: []predictLimiter{
+					{
+						predict: func(r *request.Request) bool {
+							return true
+						},
+						limiter: rate.NewLimiter(10, 5),
+					},
+					{
+						predict: func(r *request.Request) bool {
+							return false
+						},
+						limiter: rate.NewLimiter(1, 5),
+					},
+				},
+			},
+			args: args{
+				r: &request.Request{
+					HTTPRequest: &http.Request{},
+				},
+			},
+			testQPS: 100,
+			validCallsCount: func(elapsedDuration time.Duration, count int64) {
+				ideal := 5 + 10*elapsedDuration.Seconds()
+				// We should never get more requests than allowed.
+				if want := int64(ideal + 1); count > want {
+					t.Errorf("count = %d, want %d (ideal %f", count, want, ideal)
+				}
+				// We should get very close to the number of requests allowed.
+				if want := int64(ideal - 1); count < want {
+					t.Errorf("count = %d, want %d (ideal %f", count, want, ideal)
+				}
+			},
+		},
+		{
+			name: "two limiter, both matching",
```

Comment:
This test is ensuring that the burst is honored? Shouldn't the two limits be different so we can check which one is honored? Or is that non-deterministic?

Reply:
they have different rate :D. 
the purpose is to  just to check if there are different limits(rate+burst), the lower one get honered.

Reply:
Oh, I see, the fact that the lower one is 1 TPS means we'll get the burst of 5 and nothing more within 1 second. Ok cool.